src/osd/scheduler/OpSchedulerItem.h

```diff
@@ -492,17 +492,18 @@ class PGRecovery : public PGOpQueueable {
   epoch_t epoch_queued;
   uint64_t reserved_pushes;
   int priority;
+  OSDService *osd;
 public:
   PGRecovery(
     spg_t pg,
     epoch_t epoch_queued,
     uint64_t reserved_pushes,
-    int priority)
+    int priority, OSDService *service)
     : PGOpQueueable(pg),
       time_queued(ceph_clock_now()),
       epoch_queued(epoch_queued),
       reserved_pushes(reserved_pushes),
-      priority(priority) {}
+      priority(priority), osd(service) {}
```

Comment:
Explicitly disallow copy/move construction/assignment.

Reply:
Add a class comment explaining that the object releases reserved_pushes upon construction, a second comment with the same information on the reserved_pushes member, and a third on the destructor declaration.

Reply:
I will update in the next patch.

Reply:
> Explicitly disallow copy/move construction/assignment.

Can you please describe for which object you are referring? IIUC here osd is a pointer of OSDService and i am passing this pointer, it should not call any explicit constructor. Please correct me if i am wrong.

Reply:
PGOpQueueable.  The object "owns" releasing reserved_pushes, so copying it would result in two different objects releasing the same reserved_pushes.  Make that impossible by explicitly disallowing copy construction:

```
PGOpQueueable(const PGOpQueueable&)  = delete;
```

Move construction can make sense, but you'd have to implement it correctly.  Disallowing it as with copy construction would be simpler if you don't actually need it:

```
PGOpQueueable(PGOpQueueable&&)  = delete;
```

Reply:
Disallow copy and move assignment

```
PGOpQueueable &operator(const PGOpQueueable &) = delete
PGOpQueueable &operator(PGOpQueueable &&) = delete
```

for the same reason.

Reply:
done